src/coinjoin/context.h

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) 2023 The Dash Core developers
+// Distributed under the MIT/X11 software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_COINJOIN_CONTEXT_H
+#define BITCOIN_COINJOIN_CONTEXT_H
+
+#if defined(HAVE_CONFIG_H)
+#include <config/bitcoin-config.h>
+#endif
+
+#include <memory>
+
+class CBlockPolicyEstimator;
+class CChainState;
+#ifdef ENABLE_WALLET
+class CCoinJoinClientQueueManager;
+#endif // ENABLE_WALLET
+class CCoinJoinServer;
+class CConnman;
+class CMasternodeSync;
+class CTxMemPool;
+
+#ifdef ENABLE_WALLET
+struct CJClientManager;
+#endif // ENABLE_WALLET
+
+struct CJContext {
+    CJContext() = delete;
+    CJContext(const CJContext&) = delete;
+    CJContext(CChainState& chainstate, CConnman& connman, CTxMemPool& mempool, const CMasternodeSync& mn_sync, bool relay_txes);
+    ~CJContext();
+
+#ifdef ENABLE_WALLET
+    CJClientManager* const clientman;
```

Comment:
We could just make this a shared_pointer and make the global a shared pointer (they point to the same object) to ensure that we don't get dangling pointers?

Reply:
This is continuation of precedent set with `LLMQContext`'s `quorum_block_processor`, `qman`, `clhandler` and `isman` being bare pointers ([source](https://github.com/dashpay/dash/blob/13c12406ae076b58cba580aa47c3419ef5bec361/src/llmq/context.h#L51-L59)), which in turn is a continuation of precedent set by `NodeContext` ([source](https://github.com/dashpay/dash/blob/5261a4733a5edb036a856829688eb2370af02d4f/src/node/context.h#L39)).

`NodeContext` uses bare pointers to indicate they're not responsible for management of the pointers, `LLMQContext` and others, which do use the struct for initialization, use it to indicate the lack of exclusive control.

But yes, the concern of dangling pointers is present, one could reset the global pointer container, with `CJContext` having the faintest idea but switching positions would mean that `CJContext` (or anything that has access to its members) could perform a reset and code that relies on the global alias wouldn't have a clue about it. 

Both cases assume heavily erroneous code and neither case prevents the other. The switch can be done but I don't see the benefit.